spy/vm/astframe.py

```diff
@@ -803,6 +803,39 @@ def eval_expr_CmpOp(self, op: ast.CmpOp) -> W_MetaArg:
         w_opimpl = self.vm.call_OP(op.loc, w_OP, [wam_l, wam_r])
         return self.eval_opimpl(op, w_opimpl, [wam_l, wam_r])
 
+    def eval_expr_CmpChain(self, chain: ast.CmpChain) -> W_MetaArg:
+        assert chain.comparisons, "CmpChain must contain at least one comparison"
+        first_cmp = chain.comparisons[0]
+        prev_expr = first_cmp.left
+        wam_prev = self.eval_expr(prev_expr)
+        result: Optional[W_MetaArg] = None
+        saw_red = False
```

Comment:
I'm not sure to understand why you need `saw_red`.
If I understand correctly, what it does is to ensure that the final result is red if at least one of the individual operands is red. But I think this happens automatically because if you cal `eval_opimpl` with mixed-color operands, you get a red result anyway.

FWIW, I tried to comment out all the usage of `was_red` in the code and all your tests seem to pass.